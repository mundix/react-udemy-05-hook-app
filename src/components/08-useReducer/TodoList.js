import React from 'react';
import { TodoListItem } from './TodoListItem';

export const TodoList = ({ todos, handleDelete, handleToggle}) => {

    
    return (
        <ul className='list-group list-group-flush'>
            {
                todos.map((todo, index) => (
                    // TodoListITem , todo={todo} key=index, handleDelete=fn, handleToggle=fn
                    // Indice nunca como key , si no el id 
                    <TodoListItem index={todo.id} key={index} todo={todo} handleDelete={handleDelete} handleToggle={handleToggle} />
                ))
            }
        </ul>
    )
}
