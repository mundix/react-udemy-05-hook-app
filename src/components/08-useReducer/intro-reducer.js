
//Puede llamarse como querramos, pero simepre sera un stado incial
//Siempre debe ser un array de objetos
const initialState = [{ 
    id: 1,
    todo: 'Comprar Algo',
    done: false
}];

// una function corriente 
const todoReducer = (state = initialState, action) => {
    
    if(action?.type === 'agregar') {
        return [ ...state, action.payload];
    }
    //siempre debe regresar un state 
    return state;
}
// Nunca usar el array.push(), por que el push modifica o muta el objeto.
let todos = todoReducer();

const newTodo = {
    id: 2,
    todo: 'Comprar Leche',
    done: false
}

// const action = {
// puedo ponerle cualquier nomrbe
const agregarTodoAction = {
    type: 'agregar',
    payload: newTodo //este argumento sera el payload se le puede cambair el nombre, pero el mas usado es payload
}

todos = todoReducer( todos, agregarTodoAction);
console.log({todos});