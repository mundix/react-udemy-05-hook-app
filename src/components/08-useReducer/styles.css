body , html {
    padding: 60px;
}

ul * {
    background-color: #333;
    color: white;
}

.list-group-item {
    align-items: center;
    display: flex;
    justify-content: space-between;
    background-color: #333;
    color: white;
    cursor: pointer;
}

.complete {
    text-decoration: line-through;
}