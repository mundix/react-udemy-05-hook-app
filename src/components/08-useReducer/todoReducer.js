
//si no tengo valor inciial para el state, se incia con un []
export const todoReducer = (state = [], action) => {
    switch (action.type) {

        case 'add':
            return [...state, action.payload];

        case 'delete':
            return state.filter(todo => todo.id !== action.payload);

        case 'toggle':
            return state.map(todo =>
                (todo.id === action.payload)
                    ? { ...todo, done: !todo.done }
                    : todo
            );

        case 'toggle-old':
            return state.map((todo, index) => {
                // Devuelve todos los todo, si se da la condicion entonces 
                //  coge todas las propeidades del todo, pero edita el done, y deveuvle el obj todo 
                //con la propiedad modificada
                if (todo.id === action.payload) {
                    return {
                        ...todo, // todas las propiedades 
                        done: !todo.done
                    }
                } else {
                    //Caso contrario , solo devuelve el todo. 
                    return todo;
                }
            });

        default:
            // El default se  llama cuadno se inicializa 
            return state
    }
}