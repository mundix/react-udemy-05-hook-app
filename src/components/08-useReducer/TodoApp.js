import React, { useReducer } from 'react';
import { todoReducer } from './todoReducer';
import './styles.css';


const initialState = [
    {
        id: new Date().getTime(),
        desc: 'Aprender React',
        done: false
    },
    {
        id: new Date().getTime() + 1,
        desc: 'Aprender Git',
        done: false
    },
];

export const TodoApp = () => {

    // El dispatch es una function y sabe cuales funciones tienes el reducer que envio
    const [todos, dispatch] = useReducer(todoReducer, initialState);
    console.log(todos);

    const handleSubmit = (e) => {
        
        e.preventDefault();
        const newTodo = {
            id: new Date().getTime(),
            desc: 'Aprender React',
            done: false
        };

        const action = {
            type: 'add',
            payload: newTodo,
        }

        dispatch(action);

    }

    return (
        <div>
            <h1>Todo App: {todos.length}</h1>
            <hr />
            <div className="row">
                <div className="col-7">
                    <ul className='list-group list-group-flush'>
                        {
                            todos.map((todo, index) => (
                                <li key={todo.id} className='list-group-item'>
                                    <p className="text-center"> {index + 1}. {todo.desc}</p>
                                    <button className="btn btn-danger">Borrar</button>
                                </li>
                            ))
                        }
                    </ul>
                </div>
                <div className="col-5">
                    <h4>Agregar Todo</h4>
                    <hr />
                    <form onSubmit={handleSubmit}>
                        <input 
                            type="text" 
                            name='description' 
                            placeholder='Aprender...' 
                            autoComplete='off'  
                            className='form-control'
                            />
                        <button className="btn btn-outline-primary btn-block mt-2">
                            Agregar
                        </button>
                    </form>
                </div>
            </div>

        </div>
    )
}
