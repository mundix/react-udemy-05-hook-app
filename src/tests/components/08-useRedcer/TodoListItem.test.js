import React from 'react';
import { shallow } from 'enzyme';
import { TodoListItem } from '../../../components/08-useReducer/TodoListItem';
import { demoTodos } from '../../fixtures/demoTodo';


describe('Pruebas en <TodoListItem/>', () => {

    const handleDeleteMock = jest.fn();
    const handleToggleMock = jest.fn();

    let wrapper = shallow(
        <TodoListItem
            todo={demoTodos[0]}
            index={0}
            handleDelete={handleDeleteMock}
            handleToggle={handleToggleMock}
        />
    );

    test('Debe de mostarse correctamente', () => {
        expect(wrapper).toMatchSnapshot();
    })

    test('Debe llamar la funciton handelDelete', () => {
        
        //  jest.fn() 
        handleDeleteMock.mockReturnValue(1);
        wrapper.find('button').simulate('click');
        expect(handleDeleteMock).toHaveBeenCalledWith(demoTodos[0].id);
        // toHaveBeenCalled
        // toHaveBeenCalleWith
    })
    
    test('Debe llamar la funciton handelToggle', () => {
        //  jest.fn() 
        handleToggleMock.mockReturnValue(1);
        wrapper.find('p.btn-toggle-action').simulate('click');
        expect(handleToggleMock).toHaveBeenCalledWith(demoTodos[0].id);
        console.log(wrapper.html());
        expect(wrapper).toMatchSnapshot();
        // toHaveBeenCalleWith

    })
    
    test('Debe de mostrar el texto correctamente', () => {
       //contenido del parrafo 
       expect(wrapper.find('p').text().trim()).toBe(`1. ${demoTodos[0].desc}`);

    });

    test('Debe de tener la calse COMPLETE si el todo esta en TRUE', () => {
        const todo = demoTodos[0];
        todo.done = true;

        const wrapper = shallow(
            <TodoListItem
                todo={todo}
                index={0}
                handleDelete={handleDeleteMock}
                handleToggle={handleToggleMock}
            />
        );
        console.log(wrapper.html());
        expect(wrapper.find('p').hasClass('complete')).toBe(true);
    })
    
    


})
