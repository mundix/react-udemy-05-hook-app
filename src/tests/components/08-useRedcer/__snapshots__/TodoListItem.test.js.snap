// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoListItem/> Debe de mostarse correctamente 1`] = `
<li
  className="list-group-item"
  key="1"
>
  <p
    className="false btn-toggle-action"
    onClick={[Function]}
  >
    1
    . 
    Aprender React
  </p>
  <button
    className="btn btn-danger btn-delete-action"
    onClick={[Function]}
  >
    Borrar
  </button>
</li>
`;

exports[`Pruebas en <TodoListItem/> Debe llamar la funciton handelToggle 1`] = `
<li
  className="list-group-item"
  key="1"
>
  <p
    className="false btn-toggle-action"
    onClick={[Function]}
  >
    1
    . 
    Aprender React
  </p>
  <button
    className="btn btn-danger btn-delete-action"
    onClick={[Function]}
  >
    Borrar
  </button>
</li>
`;
